DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,242 @@
+        #endregion
+
+        #region ICollection Implementation
+        private int _count;
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get 
+            {
+                return _count;
+            }
+        }
+        /// <summary>
+        /// Checks whether this collection is readonly
+        /// </summary>
+        public bool IsReadOnly 
+        {
+            get 
+            {
+                return false;
+            }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            _count = 0;
+            _start = 0;
+            _end = 0;
+            _circularBuffer = new T[Length + 1];
+        }
+        /// <summary>
+        /// Checks whether the buffer contains an item
+        /// </summary>
+        public bool Contains(T item) 
+        {
+            return _circularBuffer.Contains(item);
+        }
+        /// <summary>
+        /// Copies this buffer to an array
+        /// </summary>
+        public void CopyTo(T[] array, int arrayIndex) 
+        {
+            if (array == null) 
+            {
+                throw new ArgumentNullException("array can not be null");
+            }
+
+            if (array.Length == 0 || arrayIndex >= array.Length || arrayIndex < 0) 
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            // Get enumerator
+            var enumarator = GetEnumerator();
+
+            // Copy elements if there is any in the buffer and if the index is within the valid range
+            while (arrayIndex < array.Length) 
+            {
+                if (enumarator.MoveNext()) 
+                {
+                    array[arrayIndex] = enumarator.Current;
+                    arrayIndex++;
+                }
+                else 
+                {
+                    break;
+                }
+            }
+        }
+        /// <summary>
+        /// Removes an item from the buffer
+        /// </summary>
+        public bool Remove(T item) 
+        {
+            var result = false;
+            for (int i = 0; i < _circularBuffer.Length; i++) 
+            {
+                if (item.Equals(_circularBuffer[i])) 
+                {
+                    _circularBuffer[i] = default(T);
+                    --_count;
+                    result = true;
+                    int j = i;
+                    // Moving elements one step forward
+                    while(j < Length ) 
+                    {
+                        var temp = _circularBuffer[j];
+                        _circularBuffer[j] = _circularBuffer[j + 1];
+                        _circularBuffer[j + 1] = temp;
+                        j++;
+                    }
+                    _end = i + _count ;
+                    break;
+                }
+            }
+            return result;
```

Comment:
I think the following is a much better solution for removing an element from an array:
```csharp
# Assume _circularBuffer is the following: { 1, 3, 4, 9, 2 }
if (Contains(item))
{
    _circularBuffer = _circularBuffer.Except(new T[]{ item }).ToArray();
    
    // Make sure that `_count` and `_end` remain within valid bounds 
    if (Count > 0)
    {
        --_count;
        --_end;
    }
    
    return true;
}

return false;
```

Reply:
Please add test cases for deleting the `default(T)` value from an empty buffer.

Reply:
Actually, the code I wrote above has a bug of incorrectly calculating the value for `_end` pointer. Your code also has that bug. Here's how to reproduce:

Let's say you have an almost full buffer of length 5 (last spot remains):
```
var buffer = new int[]{ 11, 12, 13, 14 }
```

At this point the `_start` pointer is equal to `0` and the `_end` pointer is equal to `4`, because: `(3 + 1) % 5 = 4`, which is the expression that gets evaluated every time you insert a new element.

What would happen if we insert a new element of value `15`? The buffer will look like:
```
{ 11, 12, 13, 14, 15}
```

And the `_end` pointer will be set to `0`, because: `(4 + 1) % 5 = 0`.

So, what would happen if we were to delete the item `14`? The buffer will look like the following:

```
{ 11, 12, 13, 15, <EMPTY> }
```

But the `_end` pointer will be set to `-1`, because the `Remove` method just decrements it. I think the `_end` pointer needs to be recalculated based on the `_count` value, right before the `_count` gets decreased, as follows:
```csharp
// Recalculate _end pointer value
_end = (Count - 1) % Length;

// Decrease the _count value now
--_count;
```

In the case of  a full buffer where `_end = _start = 0`, the `_end` pointer will be re-calculated to `4`, because: `(5 - 1) % 5 = 4`. This method scales for every size of the buffer, here are the calculations:
  * Full buffer: `_end = (5 - 1) % 5;    // 4 (5th index of the array, currently free)`
  * 4/5 full buffer: `_end = (4 - 1) % 5;    // 3 (Second-last index of the array, currently free)`
  * 3/5 full buffer: `_end = (3 - 1) % 5;    // 2 (3rd / mid index of the array, currently free)`
  * 2/5 full buffer: `_end = (2 - 1) % 5;    // 1 (2nd index of the array, currently free)`
  * 1/5 full buffer: `_end = (1 - 1) % 5;    // 0 (1st index of the array, empty buffer)`

Please review what I just wrote and let me know if anything is not clear.

Also, please document the order of operations in the new implementation, why we re-calculate the `_end` pointer first before decrementing the `_count` ... etc.

Reply:
Please add tests to cover the following:
  * Try removing the `default(T)` value from the buffer and make sure the buffer is not affected as long as the elements don't contain that value
  * Try adding duplicate entries to the buffer and making sure all of them get deleted when the `Remove(T item)` gets called and that the difference in buffer length is equal to the original length minus the number of duplicate entries that were in the buffer, originally
  * Removing elements one by one from the end of the buffer to the front, results in a correct state of the buffer and when inserting them back in the buffer looks identical to how it was before the items were deleted

Reply:
Once, I added these tests to my implementation, the bug jumped out :) . Yours is quite simple. Let me implement it and see how it goes.

Reply:
After implementing your suggested code. The first test case passed but the other two failed. I found some other bugs:

-This code  `_circularBuffer = _circularBuffer.Except(new T[]{ item }).ToArray();` shrinkes the `Length` of the buffer. The buffer is meant to be a fixed "Ring" (have fixed length) but the element count can change.

- When we remove identical elements, say we had a buffer that looks like this `{1,2,3,3,3}`, The count is only decreased by one because of this line `--_count;`, whereas it's meant to decrease by three because only two elements will be left: `{1,2}`.

- The point above also affects the `_end` pointer. Instead of decreasing by 3, it's being decreased by one because of this line `_end = (Count - 1) % Length;`.

Following your line of thought, this is what I think we should implement:
`if (!IsEmpty && Contains(item)) 
            {
               var sourceArray = _circularBuffer.Except(new T[] { item }).ToArray();
                _circularBuffer = new T[Length + 1];
                Array.Copy(sourceArray, _circularBuffer, sourceArray.Length);

                if (!item.Equals(default(T))) 
                {
                    _end = sourceArray.Length - 1;
                    _count = sourceArray.Length-1;
                }  
                return true;
            }

            return false;`
I have Implemented this on my local machine and all the test cases passed. To explain how it solved the problem, let's create a buffer with 4 elements. To the client, it looks like this: `{1,2,3,3}` but to the `CircularBuffer<T> class` it looks like this `{1,2,3,3,<empty>}`. This <empty> space is required when using two pointers as discussed in this Wikipedia article [Circular_buffer](https://en.wikipedia.org/wiki/Circular_buffer). So,

- When we remove `default(T)`, which is equivalent to `<empty>`, the `sourceArray` will look like this `{1,2,3,3}`. we now create a new `_circularBuffer`, with our initial Length. We then copy the   `sourceArray` into our empty  `_circularBuffer`. which now looks like the original - `{1,2,3,3,<empty>}`.  Since we are removing `default(T)`, or _end and _count should not change - that's what the `if` statement helps us do.

- If we remove 3 the `sourceArray` will look like this `{1,2,<empty>}`. Thus, by copying it to the newly created `_circularBuffer` we now have `{1,2,<empty>,<empty>,<empty>}`. we now have to recalculate the `_end` pointer as `_end = sourceArray.Length - 1;` (insert at index 2). The `_count` is also recalculated as `_count=sourceArray.Length-1` (equals 2).

Please review the algorithm. Once we resolve her, I'll push the changes :) . Thanks

Reply:
@NdubuisiJr, good catch! Your line of thought looks tight, please push your changes to this PR. I will mark this conversation as resolved! I am looking forward to see an extensive test coverage for this method.

Reply:
Okay here goes...